chasm/lib/activity/activity_tasks.go

```diff
@@ -64,3 +69,134 @@ func (e *activityDispatchTaskExecutor) Execute(
+	ctx chasm.Context,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	task *activitypb.ScheduleToStartTimeoutTask,
+) (bool, error) {
+	attempt, err := activity.Attempt.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if activity.Status != activitypb.ACTIVITY_EXECUTION_STATUS_SCHEDULED || task.Attempt != attempt.Count {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (e *scheduleToStartTimeoutTaskExecutor) Execute(
+	ctx chasm.MutableContext,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	_ *activitypb.ScheduleToStartTimeoutTask,
+) error {
+	return TransitionTimedOut.Apply(activity, ctx, timeoutEvent{TimeoutType: enumspb.TIMEOUT_TYPE_SCHEDULE_TO_START})
+}
+
+type scheduleToCloseTimeoutTaskExecutor struct{}
+
+func newScheduleToCloseTimeoutTaskExecutor() *scheduleToCloseTimeoutTaskExecutor {
+	return &scheduleToCloseTimeoutTaskExecutor{}
+}
+
+func (e *scheduleToCloseTimeoutTaskExecutor) Validate(
+	ctx chasm.Context,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	task *activitypb.ScheduleToCloseTimeoutTask,
+) (bool, error) {
+	attempt, err := activity.Attempt.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if !TransitionTimedOut.Possible(activity) || task.Attempt != attempt.Count {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (e *scheduleToCloseTimeoutTaskExecutor) Execute(
+	ctx chasm.MutableContext,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	_ *activitypb.ScheduleToCloseTimeoutTask,
+) error {
+	return TransitionTimedOut.Apply(activity, ctx, timeoutEvent{TimeoutType: enumspb.TIMEOUT_TYPE_SCHEDULE_TO_CLOSE})
+}
+
+type startToCloseTimeoutTaskExecutor struct{}
+
+func newStartToCloseTimeoutTaskExecutor() *startToCloseTimeoutTaskExecutor {
+	return &startToCloseTimeoutTaskExecutor{}
+}
+
+func (e *startToCloseTimeoutTaskExecutor) Validate(
+	ctx chasm.Context,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	task *activitypb.StartToCloseTimeoutTask,
+) (bool, error) {
+	attempt, err := activity.Attempt.Get(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	if activity.Status != activitypb.ACTIVITY_EXECUTION_STATUS_STARTED || task.Attempt != attempt.Count {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (e *startToCloseTimeoutTaskExecutor) Execute(
+	ctx chasm.MutableContext,
+	activity *Activity,
+	_ chasm.TaskAttributes,
+	task *activitypb.StartToCloseTimeoutTask,
+) error {
+	retryPolicy := activity.RetryPolicy
+	attempt, err := activity.Attempt.Get(ctx)
+	if err != nil {
+		return err
+	}
+
+	interval := calculateRetryInterval(retryPolicy, attempt.Count)
+	isNegativeInterval := retryPolicy.GetMaximumInterval().AsDuration() == 0 && interval <= 0 // From existing workflow activity retry behavior
+
+	// Retry task if we have remaining attempts. A retry involves transitioning the activity back to scheduled state.
+	if (retryPolicy.GetMaximumAttempts() == 0 || task.GetAttempt() < retryPolicy.GetMaximumAttempts()) && (!isNegativeInterval) {
```

Comment:
Looks like we can drop the negative interval check since this implementation doesn't return a negative value.